text/0000-service-hook.md

```diff
@@ -0,0 +1,100 @@
+* Start Date: 2019-03-21
+* RFC PR: (leave this empty)
+* Fusion Issue: (leave this empty)
+
+# Summary
+
+Add hook for consuming a registered service
+
+
+# Basic example
+
+We have a plugin registered to a token. We want to use the service that plugin provides.
+
+```javascript
+// components/Example.js
+import {ExampleToken} from 'fusion-tokens';
+import {useService} from 'fusion-react';
+
+export default function Example() {
+  const service = useService(ExampleToken);
+
+  return (
+    <button onClick={service}>Invoke service</button>
+  );
+}
+```
+
+# Motivation
+
+Consuming services through the legacy Context API is cumbersome and increasingly difficult with the adoption of hooks in Fusion. This proposal aims to provide a simplified and contemporary way of providing services to a React application.
+
+This work will not include refactoring existing plugins to hooks but will allow the migration path for all current '-react' plugins in the near future.
+
+# Detailed design
+
+The `useService` hook would get a function from Context to lookup the registered service for the supplied token.
+
+```javascript
+// fusion-react
+const ServiceContext = React.createContext();
+
+export function useService(token) {
+  const getService = useContext(ServiceContext);
+  const provides = getService(token);
+  return provides;
+}
+```
+
+The Context Provider would supply a function that can get the registered service.
+
+The naive approach is to simply provide access to the `getServices` method on the app instance. This could be done with a very simple plugin.
+
+```javascript
+// fusion-react
+export default class App extends FusionApp {
+  constructor() {
+    // ...
+    this.register(
+      createPlugin({
+        middleware() {
+          return (ctx, next) => {
+            ctx.element = ctx.element && (
+              <ServiceContext.Provider value={app.getService}>
+                {ctx.element}
+              </ServiceContext.Provider>
+            );
+            return next();
+          };
+        }
+      })
+    );
+  }
+}
+```
+
+Though I can find examples of its usage, `getServices` is an undocumented method so it's possibly intended to be private or unstable.
+
+Alternatively, the resolved plugins will need to be made available to the Context Provider in some other way. Admittedly, I would need more details here for an in-depth alternate solution. Since plugins can be overwritten, we need to be sure the DI system has resolved all plugins.
+
+# Drawbacks
+
+The simple approach is possible to do in user space (`app.js`).
```

Comment:
Not sure I understand what this line is saying

Reply:
Sorry took the term from the template. I mean that this solution could be implemented in the consumer application with a regular plugin, such as in app.js. The solution I outlined works and has no technical constraint to be in the fusion-react package. Will update